### PR TITLE
Global Styles: Site Edit Integration (FSE)

### PR DESCRIPTION
**Pull request by ItsJonQ**
> 👋 This is still a work in progress!

## Description

Screen Shot 2020-02-05 at 7 20 33 PM

This update creates an initial interface for the Global Styles controls as part of [the initial Global Styles work](https://github.com/WordPress/gutenberg/projects/40).

We're starting off with the **simplest implementation**.

The purpose is to rig up the various data/render mechanics within the Global Styles system with the FSE experience. We're currently exploring how the [user experience will be improved](https://github.com/WordPress/gutenberg/issues/19255) in regards to the controls.

## How has this been tested?

* **Enable Full Site Editing** and **Page Templates** under Gutenberg Experiments (in the left sidebar)

Screen Shot 2020-02-05 at 19 23 11

* Navigate to **Site Editor (Beta)**. Global Styles will be in a sidebar panel on the right

Screen Shot 2020-02-05 at 19 24 24

### GlobalStyles Panel

The GlobalStyles controls are rendered in a `Panel` within the new FSE `edit-site` component. This implementation is subject to change. I had added it there just to get started

### ColorControl

Part of this change involves adding a new Color picking component, `ColorControl`. We don't necessarily need to ship both the controls **and** the new global styles interface. However, this simplified control is much easier to work with compared to the other Color pickers from our current `@wordpress/components` library.

### Getting and Setting data

**This still needs to be rigged up**. For the initial push, I've just added the UI controls. We still need to resolve, hydrate, update and render the global style values

@nosolosw, @jorgefilipecosta, and I have worked together to build the various pieces.

@karmatosed has helped with initial designs.

cc'ing all you lovely folks so that we can work through this together :)

Thank you!

## Todos

* [ ] Connect data + dispatchers to get/set global style values
* [ ] Client side render global style values -> CSS on load and change
* [x] Update block CSS (e.g. Paragraph, Heading) with global style CSS variables

### CODE REVIEW

**Comment by youknowriad** [comment]
Do you think we should extract the ColorControl component to its own PR?

**Comment by karmatosed** [comment]
Really excited to see this taking shape, thanks for kicking this off @ItsJonQ

**Comment by ItsJonQ** [comment]
@youknowriad I can if that's preferable 👍 

The code originally came from my branch here:
https://github.com/WordPress/gutenberg/tree/try/color-picker-ui-update

**Comment by ItsJonQ** [comment]
Heya all! 

The latest update I've pushed includes a (mock) rigging of the global style controls with placeholder content on Edit Site. This update helps simulate the experience of setting global styles:

Screen Shot 2020-02-06 at 5 39 41 PM

**Demo GIF**
https://d.pr/i/0An0Jw

### WP GS Environment

Within the Gutenberg Editor, we can't add the `wp-gs` className onto the `html` like we can on the front-end of the site. If we do, all `h1`, `h2`, ... `p` elements will be affected. Instead, we must add it to the `div` that contains all of the rendered Gutenberg blocks. This only targets block rendered elements, rather than `h1` or `p` that may appear in the sidebar.

Note: Non-block rendered Component UI may still be affected, depending on their CSS specificity.

### Mock State/Renderer

I've written a mock state manager (`wp.data`/Redux) and a mock renderer to simulate the live rendering experience. We'll need to replace these with data/actions from `wp.data`.

### Block Components x Variables

I've rigged up the `.wp-gs` scoped variable styles with the `Paragraph` and `Heading` blocks from core.

### Placeholder content

For now, I've added `h1` -> `h6` and a `p` with lorem ipsum directly in the `front-page.html` demo template for Edit Site. This is far from ideal, but I think this is a good + simple starting point.

---

This is still **incredibly** early! It's a good starting point though!

Very cool to see all of the pieces come together :)

**Comment by ItsJonQ** [comment]
@nosolosw + @karmatosed Haiii! I have some updates :)

### By the block

I've improved the Global Styles Sidebar/Panel experience in the context of Edit Site. The biggest UX change is that block specific block styles only show when the block is selected (demo below):

Screen Capture on 2020-02-12 at 16-25-24

Note: Global site values like Color and Typography are always there

### `@wordpress/global-styles`

The new updates include a new package, `@wordpress/global-styles`. That provides the Blocks and  Editors (Post Editor and Site Editor) with the pieces they need to interface with the Global Styles system.

These mechanisms were designed to be as **agnostic as possible**. This will give was maximum flexibility as we work through changes. For example, the controls don't care where they are rendered, whether it's in the Sidebar or otherwise (cc'ing @karmatosed )

This was achieved via the `SlotFill` system from `@wordpress/components`, which is works identically to `InspectorControls`.

### Blocks x Global Style Hooks

To make the Controls block specific, I've adapted a similar API pattern as `InspectorControl`. Below is an example of a custom block using a Global style hook:

```jsx
import {
  GlobalStylesControls,
  GlobalStylesPanelBody,
  useGlobalStylesState,
} from '@wordpress/global-styles';

function MyBlockEdit( {
    attributes,
    ...etc
} ) {
  const { headingFontWeight, setHeadingFontWeight } = useGlobalStylesState();

  return (
    <>
      <GlobalStylesControls>
        <GlobalStylesPanelBody title={ __( 'MyBlock' ) }>
          <RangeControl
            label={ __( 'Font Weight' ) }
            value={ headingFontWeight }
            onChange={ setHeadingFontWeight }
            min={ 100 }
            max={ 900 }
            step={ 100 }
          />
        </GlobalStylesPanelBody>
      </GlobalStylesControls>
      <InspectorControls>
        ...
      </InspectorControls>
            ...
    </>
  );
}
```

Note: Not saying this is the definite way to go, but I think it's a good starting point!

File: packages/block-editor/src/components/index.js [hunk 1]
```diff
@@ -10,7 +10,11 @@ export { default as Autocomplete } from './autocomplete';
 export { default as BlockAlignmentToolbar } from './block-alignment-toolbar';
 export { default as BlockBreadcrumb } from './block-breadcrumb';
 export { default as BlockControls } from './block-controls';
-export { default as BlockEdit, useBlockEditContext } from './block-edit';
+export {
+	default as BlockEdit,
+	useBlockEditContext,
+	ifBlockEditSelected,
```
Comment (ItsJonQ) [inline packages/block-editor/src/components/index.js]: Need this piece for the global styles slot/fill

File: packages/global-styles/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function isEditSite() {
```
Comment (ItsJonQ) [inline packages/global-styles/src/utils.js]: Pretty hack implementation 😅

**Comment by johnstonphilip** [comment]
This is super cool so far! A couple of my initial thoughts with it:

- Is there a way we can ensure the styles being changed in this view, are actually in the view? Some of the sliders (like font scale) don't affect anything on my page, though I'm sure they'll affect something, somewhere.
- Are there plans for handling responsive global styles? That is, global styles that handle CSS media queries for certain screen widths? Certain layouts will require different things to happen at different screen widths, and will require the site owner to define those widths and the global styles that need to exist at that width. 

If these have been addressed elsewhere already, my apologies! I am doing my best to track the latest status of everything in the Red Phase!

**Comment by ItsJonQ** [comment]
@johnstonphilip Awesome! Thank you so much for checking it out and providing feedback 😍 

> Is there a way we can ensure the styles being changed in this view, are actually in the view

Not yet! At the moment, the changes aren't actually saved yet. We're working on bringing in the saving and rendering on the *actual* site part next 🙌 

> Are there plans for handling responsive global styles

It's very top of mind for me. However, the ideas I have are very premature. Handling responsiveness is something we **need to improve for blocks in general**. Changing styles from an aesthetics point of view is one thing (e.g. color), but changing values/behaviour based on layout is another. 

If you have any ideas or have seen good examples of this, let's create a Github issue and explore it together :)

> If these have been addressed elsewhere already, my apologies! I am doing my best to track the latest status of everything in the Red Phase!

No apologies needed! For global styles updates, I've tried my best to keep current work update to date on the [Github project board](https://github.com/WordPress/gutenberg/projects/40).

**Comment by mtias** [comment]
I like how this is shaping up. It's going to be easy to reason with the levels of fallback: `var(--wp-paragraph--color, var(--wp-color--text))`.

I agree with splitting the color part away, specially since we need to revise how the picker and the notion of palettes work.

It'd be interesting to connect a line-height control on the paragraph block in subsequent PRs to see how the base style relates to custom modifications.

File: packages/edit-site/src/components/editor/index.js [hunk 3]
```diff
@@ -29,21 +30,23 @@ function Editor( { settings } ) {
...
-				>
-					<Notices />
-					<Header />
-					<Sidebar />
-					<BlockEditor settings={ settings } />
-					<Popover.Slot />
-				</EntityProvider>
-			</DropZoneProvider>
-		</SlotFillProvider>
+		<GlobalStylesStateProvider>
```
Comment (mtias) [inline packages/edit-site/src/components/editor/index.js]: Does Global Style need to be the outer-most provider?

**Comment by johnstonphilip** [comment]
I just noticed that custom blocks do not appear to be registered in this `Site Edit` view. 

Screen Shot 2020-02-17 at 11 27 32 AM


If this is something that's part of the WIP and is still coming, ignore this! :)

File: packages/edit-site/src/components/editor/index.js (hunk 3 above)
Comment (ItsJonQ) [reply packages/edit-site/src/components/editor/index.js]: For this current implementation. Most likely would not once content is provided by `@wordpress/data`. At the moment, I'm simulating that data flow with a `React.Context` instance

**Comment by mapk** [comment]
> The biggest UX change is that block specific block styles only show when the block is selected (demo below):

I'll note this one is quite difficult for me to parse. It feels extremely odd to have one block selected, but see global styles change on multiple blocks (even unselected ones). I get that this is a way to surface block-specific global styles, but nevertheless, when I have one specific block selected, I only expect changes to apply to that one block. 

Global styles is... well... global. When I enter into that mode, I don't expect to have a particular block selected because I'm ready to change all related content... wherever it may live.

Where I have the most trouble is block-type global styling. I agree that it's a necessary part, and probably the hardest part for getting the UI right. I don't want to derail the work, so let's get some testing around this to see if I'm just being a cantankerous old man on this one. 😉

**Comment by karmatosed** [comment]
Just closing a loop from another issue this came up, if we can see about getting all blocks of a type selected when you click one, that might ease. See: https://github.com/WordPress/gutenberg/issues/20212#issuecomment-588268306 for context.

**Comment by ItsJonQ** [comment]
Closing this PR in favour of https://github.com/WordPress/gutenberg/pull/20530 by @nosolosw .

For context, @nosolosw forked this PR and continued work to extend it

**Pull request closed without merging**